src/dataframe/dataframe.jl

```diff
@@ -697,15 +697,24 @@ end
 ##
 ##############################################################################
 
-# hcat! for 2 arguments
+# hcat! for 2 arguments, only a vector or a data frame is allowed
 function hcat!(df1::DataFrame, df2::AbstractDataFrame)
     u = add_names(index(df1), index(df2))
     for i in 1:length(u)
         df1[u[i]] = df2[i]
     end
     return df1
 end
+
+# definition required to avoid hcat! ambiguity
+function hcat!(df1::DataFrame, df2::DataFrame)
+    invoke(hcat!, Tuple{DataFrame, AbstractDataFrame}, df1, df2)
+end
+
 hcat!(df::DataFrame, x::AbstractVector) = hcat!(df, DataFrame(Any[x]))
+hcat!(x::AbstractVector, df::DataFrame) = hcat!(DataFrame(Any[x]), df)
+hcat!(x, df::DataFrame) = throw(ArgumentError("x must be a vector or a data frame"))
```

Comment:
I'd be more precise and say `AbstractVector` or `AbstractDataFrame`.

Reply:
fixed